setup/src/Magento/Setup/Module/Di/Code/Scanner/XmlScanner.php

```diff
@@ -42,42 +44,55 @@ public function collectEntities(array $files)
             $virtualTypeQuery = "//virtualType/@name";
 
             foreach ($xpath->query($virtualTypeQuery) as $virtualNode) {
-                $virtualTypes[] = $virtualNode->nodeValue;
-            }
-
-            $regex = '/^(.*)\\\(.*)Proxy$/';
-            $query = "/config/preference[ php:functionString('preg_match', '{$regex}', @type) > 0]/@type | " .
-                "//argument[@xsi:type='object' and php:functionString('preg_match', '{$regex}', text()) > 0] |" .
-                "//item[@xsi:type='object' and php:functionString('preg_match', '{$regex}', text()) > 0] |" .
-                "/config/virtualType[ php:functionString('preg_match', '{$regex}', @type) > 0]/@type";
-            /** @var \DOMNode $node */
-            foreach ($xpath->query($query) as $node) {
-                $output[] = $node->nodeValue;
+                $virtualTypes[] = ltrim($virtualNode->nodeValue, '\\');
             }
 
-            $factoriesOutput = array_merge($factoriesOutput, $this->scanFactories($xpath));
+            $output[] = $this->scanProxies($xpath);
+            $factoriesOutput[] = $this->scanFactories($xpath);
         }
 
-        $output = array_unique($output);
-        $factoriesOutput = array_unique($factoriesOutput);
+        $output = array_unique(array_merge([], ...$output));
+        $factoriesOutput = array_unique(array_merge([], ...$factoriesOutput));
```

Comment:
Why do you extract arrays? `array_merge` expect arrays as parameters. As I see `$output` and `$factoriesOutput` are array of strings and extracting will generate fatal error on php 8.

Reply:
I updated logic above, now it contains arrays of arrays